cirq-core/cirq/transformers/dynamical_decoupling.py

```diff
@@ -149,34 +148,127 @@ def _merge_single_qubit_ops_to_phxz(
     return gate.on(q)
 
 
-def _calc_pulled_through(moment: Moment, input_pauli_ops: ops.PauliString) -> ops.PauliString:
-    """Calculates the pulled_through such that circuit(input_pauli_ops, moment.clifford_ops) is
-    equivalent to circuit(moment.clifford_ops, pulled_through).
-    """
-    clifford_ops_in_moment: list[ops.Operation] = [
-        op for op in moment.operations if _is_clifford_op(op)
-    ]
-    return input_pauli_ops.after(clifford_ops_in_moment)
+@frozen
+class _CircuitRepr:
+    class _GateType(Enum):
+        UNKOWN = 0
+        WALL_GATE = 1
+        DOOR_GATE = 2
+        INSERTABLE_GATE = 3
+
+    gate_types: dict[ops.Qid, dict[int, _CircuitRepr._GateType]]
+    need_to_stop: dict[ops.Qid, dict[int, bool]]
+    circuit: FrozenCircuit
+
+    def __init__(self, circuit: cirq.FrozenCircuit, single_qubit_gate_moments_only: bool):
```

Comment:
instead of a constructor create a class method `from_circuit` ... this way you don't need to use object.__setattr__